### PR TITLE
Add headers to snark keys

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR adds headers to the snark keys, in the format described in the RFC.

This includes PR #6801.

Querying the headers generated with `HEAD~` gives:
```
$ head -n 2 /tmp/coda_cache_dir/*
==> /tmp/coda_cache_dir/2e50a326-a975-3aca-d709-1f6e622f0b9e <==
head: error reading '/tmp/coda_cache_dir/2e50a326-a975-3aca-d709-1f6e622f0b9e': Is a directory

==> /tmp/coda_cache_dir/step-step-proving-key-blockchain-snark-step-0-90a6b80562fb4f72a3b81d44df4f005e <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"step-proving-key","identifier":"blockchain-snark-step"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length":1080458607,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"90a6b80562fb4f72a3b81d44df4f005e","identifying_hash":"90a6b80562fb4f72a3b81d44df4f005e"}

==> /tmp/coda_cache_dir/step-step-proving-key-transaction-snark-merge-1-daf93d999bbcbc6c8f9d94c531a01ad9 <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"step-proving-key","identifier":"transaction-snark-merge"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length":1077034577,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"daf93d999bbcbc6c8f9d94c531a01ad9","identifying_hash":"daf93d999bbcbc6c8f9d94c531a01ad9"}

==> /tmp/coda_cache_dir/step-step-proving-key-transaction-snark-transaction-0-997c945238b88b8ed3e9c17b7cf17850 <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"step-proving-key","identifier":"transaction-snark-transaction"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length": 272389879,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"997c945238b88b8ed3e9c17b7cf17850","identifying_hash":"997c945238b88b8ed3e9c17b7cf17850"}

==> /tmp/coda_cache_dir/vk-step-step-verification-key-blockchain-snark-step-0-90a6b80562fb4f72a3b81d44df4f005e <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"step-verification-key","identifier":"blockchain-snark-step"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length":      2120,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"90a6b80562fb4f72a3b81d44df4f005e","identifying_hash":"90a6b80562fb4f72a3b81d44df4f005e"}

==> /tmp/coda_cache_dir/vk-step-step-verification-key-transaction-snark-merge-1-daf93d999bbcbc6c8f9d94c531a01ad9 <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"step-verification-key","identifier":"transaction-snark-merge"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length":      2122,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"daf93d999bbcbc6c8f9d94c531a01ad9","identifying_hash":"daf93d999bbcbc6c8f9d94c531a01ad9"}

==> /tmp/coda_cache_dir/vk-step-step-verification-key-transaction-snark-transaction-0-997c945238b88b8ed3e9c17b7cf17850 <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"step-verification-key","identifier":"transaction-snark-transaction"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length":      2128,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"997c945238b88b8ed3e9c17b7cf17850","identifying_hash":"997c945238b88b8ed3e9c17b7cf17850"}

==> /tmp/coda_cache_dir/vk-wrap-wrap-verification-key-blockchain-snark-6d8adb26543726dbffa96b03ea051cfc <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"wrap-verification-key","identifier":"blockchain-snark"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length":      1882,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"6d8adb26543726dbffa96b03ea051cfc","identifying_hash":"6d8adb26543726dbffa96b03ea051cfc"}

==> /tmp/coda_cache_dir/vk-wrap-wrap-verification-key-transaction-snark-f868926b90bd78ccabf54737d371aed6 <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"wrap-verification-key","identifier":"transaction-snark"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length":      1890,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"f868926b90bd78ccabf54737d371aed6","identifying_hash":"f868926b90bd78ccabf54737d371aed6"}

==> /tmp/coda_cache_dir/wrap-wrap-proving-key-blockchain-snark-6d8adb26543726dbffa96b03ea051cfc <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"wrap-proving-key","identifier":"blockchain-snark"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length": 539961994,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"6d8adb26543726dbffa96b03ea051cfc","identifying_hash":"6d8adb26543726dbffa96b03ea051cfc"}

==> /tmp/coda_cache_dir/wrap-wrap-proving-key-transaction-snark-f868926b90bd78ccabf54737d371aed6 <==
MINA_SNARK_KEYS
{"header_version":1,"kind":{"type":"wrap-proving-key","identifier":"transaction-snark"},"constraint_constants":{"sub_windows_per_window":11,"ledger_depth":14,"work_delay":2,"block_window_duration_ms":180000,"transaction_capacity":{"two_to_the":7},"pending_coinbase_depth":5,"coinbase_amount":"200000000000","supercharged_coinbase_factor":2,"account_creation_fee":"1000000","fork":{}},"commits":{"mina":"ba66f83755031d7002b6a7d95ffdef45d4c0302b","marlin":"49f81edc9c86e5907d26ea791fa083640ad0ef3e"},"length": 540061195,"commit_date":"2020-11-20T19:55:09+00:00","constraint_system_hash":"f868926b90bd78ccabf54737d371aed6","identifying_hash":"f868926b90bd78ccabf54737d371aed6"}
```

Checklist:

- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

### CODE REVIEW

File: src/lib/snark_keys_header/snark_keys_header.ml [hunk 1]
```diff
@@ -300,3 +334,114 @@ let%test_module "Check parsing of header" =
...
+  (* Go to where we wrote the data *)
+  Out_channel.seek out_channel (Int64.of_int length_data_offset) ;
+  (* Pad with spaces *)
+  Out_channel.output_string out_channel true_length_padding ;
+  (* Output the true length *)
+  Out_channel.output_string out_channel true_length_string ;
+  Out_channel.close out_channel
+
+let read_with_header ~read_data filename =
+  let open Or_error.Let_syntax in
```
Comment (deepthiskumar) [inline src/lib/snark_keys_header/snark_keys_header.ml:409]: Would a unit test that does the round trip be useful here? i.e., the data and header written is the same as the data and header read? (Using dummy proof for the data)

File: src/lib/pickles/pickles.ml [hunk 2]
```diff
@@ -398,7 +419,21 @@ module Make (A : Statement_var_intf) (A_value : Statement_value_intf) = struct
...
+      { Snark_keys_header.header_version= Snark_keys_header.header_version
+      ; kind
+      ; constraint_constants
+      ; commits=
+          {mina= Coda_version.commit_id; marlin= Coda_version.marlin_commit_id}
+      ; length= (* This is a dummy, it gets filled in on read/write. *) 0
+      ; commit_date= Coda_version.commit_date
+      ; constraint_system_hash
+      ; identifying_hash=
+          (* TODO: Proper identifying hash. *)
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/pickles/pickles.ml:434]: Ideally we should be able to know whether the keys are the ones that we want without needing to generate the constraint systems that both snarks will use. I plan to use a 'representative' constraint system hash (basically compile in the one with the 'dev.mlh' constants), which won't necessarily be the hash of the particular constraint system that we are proving, but will change if the constraint system has changed.

File: src/lib/pickles/pickles.ml [hunk 3]
```diff
@@ -398,7 +419,21 @@ module Make (A : Statement_var_intf) (A_value : Statement_value_intf) = struct
...
-       ~max_branching:(module Max_branching) ~name ~typ ~choices ->
+       ~max_branching:(module Max_branching) ~name ~constraint_constants ~typ
+       ~choices ->
+    let snark_keys_header kind constraint_system_hash =
+      { Snark_keys_header.header_version= Snark_keys_header.header_version
+      ; kind
+      ; constraint_constants
+      ; commits=
+          {mina= Coda_version.commit_id; marlin= Coda_version.marlin_commit_id}
+      ; length= (* This is a dummy, it gets filled in on read/write. *) 0
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/pickles/pickles.ml:430]: It seemed cleaner to add a dummy value pre-write-time than to handle an option everywhere post-read-time. I would prefer not to change this, unless you can see it causing problems further down the line?

File: src/lib/snark_keys_header/snark_keys_header.ml (hunk 1 above)
Comment (mrmr1993) [reply src/lib/snark_keys_header/snark_keys_header.ml:409]: Anything that depends on snark keys in CI will break if it doesn't round-trip, so it should be pretty obvious :grin: More seriously, the existing unit tests check the majority of the round-trip; the remaining untested parts are `read_data`/`append_data` being proper inverses and the behaviour of multiple writes + seeking, both of which we can hopefully take as givens.

**Review by deepthiskumar — APPROVED** [review]

**Pull request merged**